# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/WindowContext.h
# Commit: 812ed5275f0d
# Full Hash: 812ed5275f0db9f92944955f891df2ab68c9f8e7
# Author: Toshihito Kikuchi <tkikuchi@mozilla.com>
# Date: 2021-11-19 04:11:43
# Regressor Bug: 1734099
# File Overlap Count: 5
# Description:
#   Bug 1734099 - Never unload tabs with active RTCPeerConnection instances.  r=peterv,jesup
#   
#   This patch makes sure the Tab Unloading feature does not unload tabs that have
#   active peer connections not to disrupt browsing experience based on WebRTC peer
#   connections.
# ==============================================================================

diff -r a733d049dbda -r 812ed5275f0d docshell/base/WindowContext.h
--- a/docshell/base/WindowContext.h	Fri Nov 19 01:03:56 2021 +0200
+++ b/docshell/base/WindowContext.h	Thu Nov 18 22:59:12 2021 +0000
@@ -87,6 +87,9 @@
   /* Whether the corresponding document has `loading='lazy'`             \
    * images; It won't become false if the image becomes non-lazy */      \
   FIELD(HadLazyLoadImage, bool)                                          \
+  /* Whether any of the windows in the subtree rooted at this window has \
+   * active peer connections or not (only set on the top window). */     \
+  FIELD(HasActivePeerConnections, bool)                                  \
   /* Whether we can execute scripts in this WindowContext. Has no effect \
    * unless scripts are also allowed in the BrowsingContext. */          \
   FIELD(AllowJavascript, bool)                                           \
@@ -194,6 +197,8 @@
 
   bool HadLazyLoadImage() const { return GetHadLazyLoadImage(); }
 
+  bool HasActivePeerConnections();
+
   bool AllowJavascript() const { return GetAllowJavascript(); }
   bool CanExecuteScripts() const { return mCanExecuteScripts; }
 
@@ -290,6 +295,8 @@
               ContentParent* aSource);
   void DidSet(FieldIndex<IDX_AllowJavascript>, bool aOldValue);
 
+  bool CanSet(FieldIndex<IDX_HasActivePeerConnections>, bool, ContentParent*);
+
   void DidSet(FieldIndex<IDX_HasReportedShadowDOMUsage>, bool aOldValue);
 
   void DidSet(FieldIndex<IDX_SHEntryHasUserInteraction>, bool aOldValue);